### PR TITLE
Markdown preview syntax highlighting

### PR DESCRIPTION
**Pull request by AndyScherzinger**
adds syntax highlighting capabilities within the markdown rendering

to be merged _after_ #4692

extends #3302 by adding syntax highlighting

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -5,4 +5,4 @@ NC_TEST_SERVER_PASSWORD=test
 android.enableJetifier=true
 android.useAndroidX=true
 android.debug.obsoleteApi=true
-
+org.gradle.jvmargs=-Xmx4g -XX:MaxPermSize=2048m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
```
Comment (tobiasKaminsky) [inline gradle.properties]: This should not be in public grade.properties, but only in your local one.

**Comment by AndyScherzinger** [comment]
Hi @stefan-niedermann,

in generall, yes, that would be the prefered way. In this case the situation is slightly diffrent, so IMHO we rather won't move for such a solution due to the next iterations of this feature.

Since Nextcloud now comes with Nextcoud Text, the client will implement support for this (similar to the collabora support). So Markdown suppurt read/write will come with supporting Nextcloud Text (end of this year, beginning of next year). So this is a rather temporary solution to suppoort markdown previewing for the time being. Other than that the advantage of Markwon (for preview-only scenrarios) is that is works natively as in rendering spannedText based on the  markdown.

So _if_ Nextcloud Text wouldn't be there I'd also vouched for using Markor 👍

File: gradle.properties (hunk 1 above)
Comment (AndyScherzinger) [reply gradle.properties]: removed 👍

**Comment by stefan-niedermann** [comment]
Did i understand you right, that the Files app will support editing markdown files in tge future based on a webview-solution (instead of a native implememtation)?

**Comment by AndyScherzinger** [comment]
> support editing markdown files in the future based on a webview-solution (instead of a native implememtation)

That is what I understood from a discussion with @tobiasKaminsky while it'll rather be a _"any-text-based"_ file editing. It is than based on a webview _and_ will of course online work when being online.

**Comment by stefan-niedermann** [comment]
Thank you for clarification :)

File: spotbugs-filter.xml [hunk 2]
```diff
@@ -33,6 +33,11 @@
             <Class name="~.*\.DataBinderMapperImpl"/>
         </Or>
     </Match>
+	
+	<!-- Third-party library code -->
+	<Match>
+	    <Package name="~io\.noties\..*" />
```
Comment (AndyScherzinger) [inline spotbugs-filter.xml]: @tobiasKaminsky mind having a look? I just don't get this exclusion right :/

Comment (AndyScherzinger) [reply spotbugs-filter.xml]: @tobiasKaminsky nevermind, got it now :)

File: build.gradle [hunk 3]
```diff
@@ -327,6 +329,10 @@ dependencies {
     implementation ("io.noties.markwon:html:$markwonVersion")
     implementation ("io.noties.markwon:linkify:$markwonVersion")
 
+    implementation ("io.noties.markwon:syntax-highlight:$markwonVersion")
+    implementation ("io.noties:prism4j:${prismVersion}")
```
Comment (tobiasKaminsky) [inline build.gradle]: Just for consistent reasons, I guess we do not need { } here, or?

File: build.gradle [hunk 4]
```diff
@@ -327,6 +329,10 @@ dependencies {
...
+    kapt "io.noties:prism4j-bundler:${prismVersion}"
```
Comment (tobiasKaminsky) [inline build.gradle]: Same here

**Comment by tobiasKaminsky** [comment]
@AndyScherzinger 
tested and works :tada: 
Can you have a look at my two (small) questions?

File: build.gradle (hunk 4 above)
Comment (AndyScherzinger) [reply build.gradle]: Done https://github.com/nextcloud/android/pull/4701/commits/f5ebba9efffef00912302fd493538767da424de0

**Comment by tobiasKaminsky** [comment]
Two spotbugs are this:


  | com.owncloud.android.ui.preview.MarkwonGrammarLocator$1.name() may return null, but is declared @Nonnull
-- | --
  | Bug type NP_NONNULL_RETURN_VIOLATION (click for details) In class com.owncloud.android.ui.preview.MarkwonGrammarLocator$1In method com.owncloud.android.ui.preview.MarkwonGrammarLocator$1.name()At MarkwonGrammarLocator.java:[line 47]
NP | com.owncloud.android.ui.preview.MarkwonGrammarLocator$1.tokens() may return null, but is declared @Nonnull
  | Bug type NP_NONNULL_RETURN_VIOLATION (click for details) In class com.owncloud.android.ui.preview.MarkwonGrammarLocator$1In method com.owncloud.android.ui.preview.MarkwonGrammarLocator$1.tokens()At MarkwonGrammarLocator.java:[line 53]


There is nothing we can do, so I will increase spotbugs.

**Comment by AndyScherzinger** [comment]
Yeah, I tried to put the file on the exclude list but failed to get the regex right 😞

**Pull request merged**